pkg/verifier/vulnerability/ibm_va.go

```diff
@@ -0,0 +1,264 @@
+func (s *IBMVAScanner) CanImageDeployBasedOnVulnerabilities(image image.Reference) (scan ScanResponse, err error) {
+	if !image.HasIBMRepo() {
+		return scan, fmt.Errorf("Cannot use IBM Vulnerability Advisor with image %q", image.String())
+	}
+	var summary IBMVASummary
+	summary, err = s.getImageStatus(image)
+	if err == nil {
+		switch summary.Status {
+		// WARN is returned when exemptions cover all of the vulnerabilities in the image
+		case "OK", "WARN", "UNSUPPORTED":
+			glog.Infof("Image %s CAN DEPLOY with IBM VA status %q", image.String(), summary.Status)
+			scan.CanDeploy = true
+		default:
+			reason := fmt.Sprintf("Image %s CANNOT DEPLOY with IBM VA status %q", image.String(), summary.Status)
+			glog.Info(reason)
+			scan.DenyReason = reason
+		}
+	} else {
+		reason := fmt.Sprintf("Image %s CANNOT DEPLOY due to IBM VA error: %q", image.String(), err)
+		glog.Info(reason)
+		scan.DenyReason = reason
+	}
+	return
+}
+
+// internal call to VA - API docs: https://cloud.ibm.com/apidocs/container-registry/va#imagestatusquerypath
+// GET /va/api/v3/report/image/status/{name}
+func (s *IBMVAScanner) getImageStatus(image image.Reference) (IBMVASummary, error) {
+	if len(s.credentials) == 0 {
+		return IBMVASummary{}, fmt.Errorf("No credentials on client to call VA with")
+	}
+
+	uri := fmt.Sprintf("https://%s/va/api/v3/report/image/status/%s", image.GetHostname(), image.String())
+
+	var summary *IBMVASummary
+	for _, cred := range s.credentials {
+		sum, err := s.callVA(cred, uri)
+		switch {
+		case err == nil:
+			break
+		case errors.Is(err, ErrorUnauthorised):
+			continue
+		default:
+			return IBMVASummary{}, fmt.Errorf("Failed to get IBM vulnerability advisor scan result for %q: %v", image.String(), err)
+		}
+
+		if sum != nil {
+			summary = sum
+			break
+		}
+	}
+
+	if summary == nil {
+		return IBMVASummary{}, fmt.Errorf("Not authorised to get IBM vulnerability scan result for %q: %w", image.String(), ErrorUnauthorised)
+	}
+
+	return *summary, nil
+}
+
+func (s *IBMVAScanner) callVA(cred credential.Credential, uri string) (*IBMVASummary, error) {
+	req, err := s.createRequest(cred, uri)
+	if err != nil {
+		return nil, err
+	}
+
+	for try := 0; try < maxRetries; try++ {
+		// Exponential backoff
+		time.Sleep(sleepTime * time.Duration(try))
+
+		var resp *http.Response
+		resp, err = s.client.Do(req)
+		if err != nil {
+			continue
+		}
+
+		defer resp.Body.Close()
+
+		switch resp.StatusCode {
+		case http.StatusOK:
+			break
+		case http.StatusUnauthorized, http.StatusForbidden:
+			return nil, ErrorUnauthorised
+		case http.StatusNotFound:
+			return nil, errors.New("Not found")
+		case http.StatusBadRequest:
+			return nil, errors.New("Bad request")
+		case http.StatusBadGateway, http.StatusInternalServerError, http.StatusServiceUnavailable:
+			err = errors.New("Internal server error")
+			continue
+		default:
+			err = errors.New("Unhandled response from IBM Vulnerability Advisor")
```

Comment:
We should log what the unexpected response code was, even if we don't return it in the error object.